[RUBY-2704] Fix bug where refund can be higher than balance

jjromeo opened the pull request:
- Enhanced validation to check if refund amount is greater than or equal to the balance.
- Removed redundant balance assignment in test setup.
- Added a test case to ensure refund amount does not exceed balance.
- Refactor error messages to use the convention for displaying from en.yml
- set date in payments table to date_time to fix regression in table
- Modified the refund form to display dynamic payment types based on `@payment.payment_type`.
- Refactored form fields to use `govuk_text_field` and `govuk_text_area` to fix issue with errors not linking back to ifelds
- Updated locale file to include additional payment types and improve translation keys.

jjromeo committed:
[RUBY-2704] Fix bug where refund can be higher than balance

- Enhanced validation to check if refund amount is greater than or equal to the balance.
- Removed redundant balance assignment in test setup.
- Added a test case to ensure refund amount does not exceed balance.

jjromeo committed:
[RUBY-2704] Refactor `RecordRefundForm` validation logic

- Simplified validation messages by removing inline I18n translations and using error symbols.
- Removed `reason_present_in_comments` method and replaced it with a direct presence validation on `comments`.
- Updated the locale file to reflect changes in validation message structure, moving messages to `activemodel.errors` format for better consistency and maintainability.

jjromeo committed:
[RUBY-2704] set date in payments table to date_time to fix issue in table

jjromeo committed:
Merge branch 'main' into RUBY-2704-wex-charging-bo-3-1-3-refund-a-payment-page

jjromeo committed:
[RUBY-2704] Rubocop fix

- reduce safe navigation chain

jjromeo committed:
[RUBY-2704] Refactor `RecordRefundForm` to improve nil handling

- Extracted `account` from `payment` to safely handle potential nil values before accessing `balance`.

jjromeo committed:
[RUBY-2704] Refactor `RecordRefundForm` to improve payment validation

- Removed `payment_exists` validation method and replaced it with `check_payment_exists?` method, which returns a boolean.
- Ensured `check_payment_exists?` is called before accessing `payment` to prevent errors.
- Simplified balance calculation by removing unnecessary checks and directly dividing `payment.account.balance` by 100.

jjromeo committed:
[RUBY-2704] Update refund page to use dynamic payment type and refactor form fields

- Modified the refund form to display dynamic payment types based on `@payment.payment_type`.
- Refactored form fields to use `govuk_text_field` and `govuk_text_area` to fix issue with errors not linking back to ifelds
- Updated locale file to include additional payment types and improve translation keys.

jjromeo committed:
[RUBY-2704] Update `waste_exemptions_engine` gem to latest revision in `Gemfile.lock`

brujeo commented on app/forms/record_refund_form.rb:
```diff
@@ -7,19 +7,21 @@ class RecordRefundForm
     @payment = WasteExemptionsEngine::Payment.find_by(id: payment_id)
-    @balance = payment&.account&.balance
+
+    @balance = payment.account.balance / 100
```
It would make sense to use existing helper for price conversion
display_pence_as_pounds_and_cents
or maybe even add a convenience method 
`payment.account.balance_in_pounds`

brujeo commented on app/forms/record_refund_form.rb:
```diff
@@ -36,28 +38,20 @@ def submit(params, record_refund_service: RecordRefundService)
-    if amount.to_f > payment.payment_amount.to_f / 100
-      errors.add(:amount, I18n.t(".record_refunds.form.errors.exceeds_payment_amount"))
-    end
+    errors.add(:amount, :exceeds_payment_amount) if amount.to_f > payment.payment_amount.to_f / 100
```
same here,  consider using existing helper method for price conversion
`display_pence_as_pounds_and_cents `

jjromeo committed:
[RUBY-2704] Convert refund amounts from pounds to pence

- Updated `RecordRefundForm` to convert refund amounts from pounds to pence using `CurrencyConversionService`.
- Modified `RecordRefundService` to handle amounts in pence instead of pounds.
- Adjusted related tests in `record_refund_service_spec.rb` to use amounts in pence.

jjromeo committed:
[RUBY-2704] Remove division by 100 for balance and payment amount checks

Adjusted the `@balance` and `amount_within_limits` methods in `RecordRefundForm` to use the raw balance and payment amount values without dividing by 100. This ensures that the calculations are accurate and consistent with the stored values.

jjromeo committed:
[RUBY-2704] Refactor `RecordRefundForm` to improve amount handling

- Removed direct conversion of amount to pence in `submit` method.
- Introduced `amount_in_pence` method for consistent conversion and validation.
- Simplified validation logic for refund amount limits.

jjromeo committed:
[RUBY-2704] Refactor `RecordRefundForm` to directly use `RecordRefundService`

Removed the dependency injection of `RecordRefundService` in the `submit` method of `RecordRefundForm` to simplify the code by directly calling `RecordRefundService.run`.

jjromeo committed:
[RUBY-2704] Update balance calculation in `PaymentDetailsPresenter` spec

- Refactored balance expectation to use `WasteExemptionsEngine::CurrencyConversionService.convert_pence_to_pounds` for converting account balance from pence to pounds.

jjromeo committed:
[RUBY-2704] Fix indentation in `payment_details_presenter_spec.rb` for balance expectation test.

brujeo commented on app/forms/record_refund_form.rb:
```diff
@@ -7,57 +7,55 @@ class RecordRefundForm
 
     return false unless valid?
 
     Rails.logger.info "running RecordRefundService with arguments: #{comments}, #{payment}, #{amount.to_f}"
```
Is this log mesage still needed? This log message looks pretty useless without the context: account_id, payment_id, etc

jjromeo committed:
[RUBY-2704] Remove logging statement from `RecordRefundForm` in preparation for production use.
